src/context.rs

```diff
@@ -1,167 +1,97 @@
-use crate::ffi;
-use crate::selinux_macros::{wrap_context_get, wrap_context_set};
-use std::ffi::{CStr, CString};
-use std::fmt;
-
-pub struct Context(ffi::context_t);
+use std::fmt::{Debug, Display, Formatter};
+pub struct Context {
+    user: String,
+    role: String,
+    _type: String,
+    range: String,
+}
 
-impl Context {
-    pub fn new(s: &str) -> Option<Self> {
-        let cs = CString::new(s).ok()?;
+macro_rules! context_access {
+    ($field:ident, $setter:ident) => {
+        pub fn $field(&self) -> &str {
+            self.$field.as_str()
+        }
 
-        match unsafe { ffi::context_new(cs.as_ptr() as *const i8) } {
-            p if !p.is_null() => Some(Self { 0: p }),
-            _ => None,
+        pub fn $setter(&mut self, $field: &str) {
+            self.$field = $field.to_owned();
         }
-    }
+    };
+}
 
-    pub fn to_str(&self) -> Option<&str> {
-        unsafe {
-            match ffi::context_str(self.0) {
-                p if !p.is_null() => CStr::from_ptr(p).to_str().ok(),
-                _ => None,
-            }
+impl Context {
+    pub fn new(context: &str) -> Option<Self> {
+        let mut iter = context.split(":");
+        let user = iter.next()?.to_owned();
+        let role = iter.next()?.to_owned();
+        let _type = iter.next()?.to_owned();
+        let range = iter.collect::<Vec<&str>>();
+        if range.len() == 0 {
+            return None;
```

Comment:
this may be incompatible with context_new

Reply:
context_new("system_u:object_r:bin_t") will return a valid context_t and you can use context_range_set to set range later.

Reply:
can test with:
```c
#include <selinux/selinux.h>
#include <selinux/context.h>
#include <stdio.h>
#include <assert.h>

int main()
{
        context_t con;
        int ret;
        char *s;

//      con = context_new("system_u:");
        con = context_new("system_u:object_r:bin_t");
        assert(con != NULL);
        //con= context_new("system_u:object_r:bin_t:s0");

        printf("user: %s\n", context_user_get(con));
        printf("role: %s\n", context_role_get(con));
        printf("type: %s\n", context_type_get(con));
        printf("range: %s\n", context_range_get(con));

        context_range_set(con, "s0");
        printf("range: %s\n", context_range_get(con));

        s = context_str(con);

        ret = security_check_context(s);

        printf("ret=%d\n", ret);

        context_free(con);
        return 0;
}
```

Reply:
Is `range` the only field that can be optional?

Reply:
Issue addressed.